File: ios/Classes/BleScanner.swift
import Foundation
import CoreBluetooth

/// Handles BLE scanning for nearby mesh devices
class BleScanner: NSObject {
    private let tag = "BleScanner"

    private var centralManager: CBCentralManager?
    private var isScanning = false
    private var scanTimer: Timer?

    // Callbacks
    var onDeviceDiscovered: ((Peer) -> Void)?
    var onScanError: ((Int, String) -> Void)?

    override init() {
        super.init()
    }

    /// Initialize the central manager
    func initialize() {
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    /// Get the central manager
    func getCentralManager() -> CBCentralManager? {
        return centralManager
    }

    /// Start scanning for BLE mesh devices
    func startScanning() {
        guard let centralManager = centralManager else {
            print("[\(tag)] Central manager not initialized")
            onScanError?(-1, "Central manager not initialized")
            return
        }

        if isScanning {
            print("[\(tag)] Already scanning")
            return
        }

        guard centralManager.state == .poweredOn else {
            print("[\(tag)] Bluetooth is not powered on")
            onScanError?(-1, "Bluetooth is not powered on")
            return
        }

        // Start scanning for mesh service UUID
        let serviceUUIDs = [BleConstants.meshServiceUUID]
        let options: [String: Any] = [
            CBCentralManagerScanOptionAllowDuplicatesKey: true
        ]

        centralManager.scanForPeripherals(withServices: serviceUUIDs, options: options)
        isScanning = true
        print("[\(tag)] Started BLE scanning")

        // Schedule scan stop after scan period
        scanTimer?.invalidate()
        scanTimer = Timer.scheduledTimer(withTimeInterval: BleConstants.scanPeriodMs, repeats: false) { [weak self] _ in
            self?.stopScanning()
        }
    }

    /// Stop scanning for BLE devices
    func stopScanning() {
        guard let centralManager = centralManager else {
            return
        }

        if !isScanning {
            return
        }

        centralManager.stopScan()
        isScanning = false
        scanTimer?.invalidate()
        scanTimer = nil
        print("[\(tag)] Stopped BLE scanning")
    }

    /// Check if currently scanning
    func isScanningActive() -> Bool {
        return isScanning
    }

    /// Clean up resources
    func cleanup() {
        stopScanning()
        centralManager = nil
        onDeviceDiscovered = nil
        onScanError = nil
    }
}

// MARK: - CBCentralManagerDelegate
extension BleScanner: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            print("[\(tag)] Bluetooth powered on")
        case .poweredOff:
            print("[\(tag)] Bluetooth powered off")
            if isScanning {
                stopScanning()
            }
            onScanError?(-1, "Bluetooth powered off")
        case .unauthorized:
            print("[\(tag)] Bluetooth unauthorized")
            onScanError?(-1, "Bluetooth unauthorized")
        case .unsupported:
            print("[\(tag)] Bluetooth unsupported")
            onScanError?(-1, "Bluetooth unsupported")
        case .resetting:
            print("[\(tag)] Bluetooth resetting")
        case .unknown:
            print("[\(tag)] Bluetooth state unknown")
        @unknown default:
            print("[\(tag)] Bluetooth state unknown")
        }
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        let rssiValue = RSSI.intValue

        // Filter out devices with very weak signal
        guard rssiValue > -100 else {
            return
        }

        print("[\(tag)] Discovered device: \(peripheral.identifier.uuidString), RSSI: \(rssiValue)")

        // Create peer from peripheral
        let peer = Peer.fromPeripheral(peripheral, rssi: rssiValue)

        // Notify callback
        onDeviceDiscovered?(peer)
    }

    // Connection callbacks are now handled by BleConnectionManager
}

File: android/src/main/kotlin/com/ble_mesh/BleScanner.kt
package com.ble_mesh

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.util.Log
import com.ble_mesh.models.Peer

/**
 * Handles BLE scanning for nearby mesh devices
 */
class BleScanner(private val context: Context) {
    private val tag = "BleScanner"

    private val bluetoothManager: BluetoothManager =
        context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager

    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager.adapter
    private val bluetoothLeScanner: BluetoothLeScanner? = bluetoothAdapter?.bluetoothLeScanner

    private val handler = Handler(Looper.getMainLooper())
    private var isScanning = false
    private var scanCallback: ScanCallback? = null

    // Callback for scan results
    var onDeviceDiscovered: ((Peer) -> Unit)? = null
    var onScanError: ((Int, String) -> Unit)? = null

    /**
     * Start scanning for BLE mesh devices
     */
    fun startScanning() {
        if (isScanning) {
            Log.d(tag, "Already scanning")
            return
        }

        if (bluetoothAdapter == null || !bluetoothAdapter.isEnabled) {
            Log.e(tag, "Bluetooth is not available or not enabled")
            onScanError?.invoke(-1, "Bluetooth is not available or not enabled")
            return
        }

        if (bluetoothLeScanner == null) {
            Log.e(tag, "BLE scanner is not available")
            onScanError?.invoke(-1, "BLE scanner is not available")
            return
        }

        // Create scan filters to only find mesh devices
        val scanFilters = listOf(
            ScanFilter.Builder()
                .setServiceUuid(ParcelUuid(BleConstants.MESH_SERVICE_UUID))
                .build()
        )

        // Create scan settings
        val scanSettings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .setCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES)
            .setMatchMode(ScanSettings.MATCH_MODE_AGGRESSIVE)
            .setNumOfMatches(ScanSettings.MATCH_NUM_MAX_ADVERTISEMENT)
            .setReportDelay(0)
            .build()

        // Create scan callback
        scanCallback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                super.onScanResult(callbackType, result)
                handleScanResult(result)
            }

            override fun onBatchScanResults(results: List<ScanResult>) {
                super.onBatchScanResults(results)
                results.forEach { handleScanResult(it) }
            }

            override fun onScanFailed(errorCode: Int) {
                super.onScanFailed(errorCode)
                Log.e(tag, "Scan failed with error code: $errorCode")
                isScanning = false
                onScanError?.invoke(errorCode, "Scan failed with error code: $errorCode")
            }
        }

        try {
            bluetoothLeScanner.startScan(scanFilters, scanSettings, scanCallback)
            isScanning = true
            Log.d(tag, "Started BLE scanning")

            // Schedule scan stop after SCAN_PERIOD_MS
            handler.postDelayed({
                stopScanning()
            }, BleConstants.SCAN_PERIOD_MS)
        } catch (e: SecurityException) {
            Log.e(tag, "Security exception while starting scan", e)
            onScanError?.invoke(-1, "Security exception: ${e.message}")
        } catch (e: Exception) {
            Log.e(tag, "Exception while starting scan", e)
            onScanError?.invoke(-1, "Exception: ${e.message}")
        }
    }

    /**
     * Stop scanning for BLE devices
     */
    fun stopScanning() {
        if (!isScanning) {
            return
        }

        try {
            scanCallback?.let {
                bluetoothLeScanner?.stopScan(it)
                Log.d(tag, "Stopped BLE scanning")
            }
        } catch (e: SecurityException) {
            Log.e(tag, "Security exception while stopping scan", e)
        } catch (e: Exception) {
            Log.e(tag, "Exception while stopping scan", e)
        } finally {
            isScanning = false
            scanCallback = null
        }
    }

    /**
     * Handle a scan result
     */
    private fun handleScanResult(result: ScanResult) {
        val device = result.device
        val rssi = result.rssi

        // Extract senderId from service data
        val serviceData = result.scanRecord?.getServiceData(
            ParcelUuid(BleConstants.MESH_SERVICE_UUID)
        )

        val senderId = if (serviceData != null && serviceData.size >= 6) {
            // Convert first 6 bytes to senderId string
            DeviceIdManager.compactIdToString(serviceData.sliceArray(0..5))
        } else {
            null  // Will be obtained via handshake after connection
        }

        if (senderId != null) {
            Log.d(tag, "Discovered device: ${device.address}, RSSI: $rssi, senderId: $senderId")
        } else {
            Log.d(tag, "Discovered device: ${device.address}, RSSI: $rssi (senderId not in advertisement)")
        }

        // Create peer from device with senderId
        val peer = Peer.fromDevice(device, rssi, senderId)

        // Notify callback
        onDeviceDiscovered?.invoke(peer)
    }

    /**
     * Check if currently scanning
     */
    fun isScanning(): Boolean = isScanning

    /**
     * Clean up resources
     */
    fun cleanup() {
        stopScanning()
        handler.removeCallbacksAndMessages(null)
        onDeviceDiscovered = null
        onScanError = null
    }
}

